### PR TITLE
🎉Source Twilio: New Stream Conversation Participants

### LINKED ISSUE #19740: Source Twilio: new stream Conversation Participants
## Tell us about the problem you're trying to solve
GET https://conversations.twilio.com/v1/Conversations/{ConversationSid}/Participants

## Describe the solution you’d like
Research pros and cons for implementing this stream

### PR DESCRIPTION
**Pull request by btkcodedev**
## What
Closes #19740
GET [https://conversations.twilio.com/v1/Conversations/{ConversationSid}/Participants](https://conversations.twilio.com/v1/Conversations/{ConversationSid}/Participants)

## Enhancement
New streams: 

- Conversation Participants

## Recommended reading order
1. `source_twilio/source.py`
2. `source_twilio/streams.py`
3.  `source_twilio/schemas/conversation_participants.json`

## 🚨 User Impact 🚨
No breaking changes

## Pre-merge Checklist
<details open><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter [Yes]

- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [x] Issue acceptance criteria met
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [x] Documentation which references the generator is updated as needed

</details>

## For the reviewer: 
- Need help with incremental streams, Please someone review with comments about how it can be improved. It would greatly help
 - Enhancement passing in local python integration tests (Except incremental). Needs test deployment to cloud build and publish it :)

### CODE REVIEW

**Comment by YowanR** [comment]
@marcosmarxm @sajarin Can you take a look, please?

File: airbyte-integrations/connectors/source-twilio/integration_tests/configured_catalog.json [hunk 1]
```diff
@@ -0,0 +1,172 @@
+{
+    "streams": [
+      {
+        "stream": {
+          "name": "addresses",
+          "json_schema": {},
+          "supported_sync_modes": ["full_refresh"]
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-twilio/integration_tests/configured_catalog.json:7]: Why are you adding this file? Is it going to be used anywhere?

File: airbyte-integrations/connectors/source-twilio/source_twilio/schemas/conversation_participants.json [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            "last_read_message_index",
+            "last_read_timestamp"
+          ]
+        }
+      }
+    },
+    "required": [
+      "participants"
+    ]
+  }
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-twilio/source_twilio/schemas/conversation_participants.json:86]: EOF and please format the json file.

File: airbyte-integrations/connectors/source-twilio/source_twilio/streams.py [hunk 3]
```diff
@@ -363,6 +364,24 @@ class Keys(TwilioNestedStream):
...
+    # uri_from_subresource = False
+
+    def path(self, stream_slice: Mapping[str, Any], **kwargs):
+        return f"Conversations/{stream_slice['conversation_sid']}/Participants"
+
+    def parent_record_to_stream_slice(self, record: Mapping[str, Any]) -> Mapping[str, Any]:
+        if(record["conversation_sid"]):
+            return {"conversation_sid": record["conversation_sid"]}
+        else:
+            return True
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-twilio/source_twilio/streams.py:384]: This is not correct. Look others implementations of this function.

File: airbyte-integrations/connectors/source-twilio/source_twilio/schemas/conversation_participants.json [hunk 4]
```diff
@@ -0,0 +1,86 @@
+{
+    "$schema": "http://json-schema.org/draft-07/schema#",
+    "title": "Generated schema for Root",
+    "type": "object",
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-twilio/source_twilio/schemas/conversation_participants.json:4]: Add `  "additionalProperties": true`

File: airbyte-integrations/connectors/source-twilio/source_twilio/schemas/conversation_participants.json [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    "properties": {
+      "participants": {
+        "type": ["null","array"],
+        "items": {
+          "type": "object",
+          "properties": {
+            "account_sid": {
+              "type": "string"
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-twilio/source_twilio/schemas/conversation_participants.json:12]: usually it is used `["null", "string"]` to not break the tests if the field is not required or is empty.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
@btkcodedev I left some comments.

File: airbyte-integrations/connectors/source-twilio/source_twilio/schemas/conversation_participants.json (hunk 4 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-twilio/source_twilio/schemas/conversation_participants.json]: Done

File: airbyte-integrations/connectors/source-twilio/source_twilio/schemas/conversation_participants.json [hunk 6]
```diff
@@ -0,0 +1,79 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "Generated schema for Root",
+  "type": "object",
+  "additionalProperties": true,
+  "properties": {
+    "participants": {
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-twilio/source_twilio/schemas/conversation_participants.json:7]: Hello @btkcodedev you want t retrieve the individual records from participants AND not the array. The schema must be the data inside the `participants`object.

File: airbyte-integrations/connectors/source-twilio/source_twilio/schemas/conversation_participants.json [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+          "date_created",
+          "date_updated",
+          "url",
+          "last_read_message_index",
+          "last_read_timestamp"
+        ]
+      }
+    }
+  },
+  "required": ["participants"]
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-twilio/source_twilio/schemas/conversation_participants.json:78]: Remove this if not strictly required.

File: airbyte-integrations/connectors/source-twilio/source_twilio/schemas/conversation_participants.json [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+            "type": ["null", "string"]
+          },
+          "date_updated": {
+            "type": ["null", "string"]
+          },
+          "url": {
+            "type": ["null", "string"]
+          },
+          "last_read_message_index": {},
+          "last_read_timestamp": {}
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-twilio/source_twilio/schemas/conversation_participants.json:61]: That type are this data?

File: airbyte-integrations/connectors/source-twilio/source_twilio/schemas/conversation_participants.json (hunk 7 above)
Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-twilio/source_twilio/schemas/conversation_participants.json:78]: If the data can be null this won't be needed.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
Some comments tests are not passing because the schema is wrong.

File: airbyte-integrations/connectors/source-twilio/source_twilio/schemas/conversation_participants.json (hunk 6 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-twilio/source_twilio/schemas/conversation_participants.json]: Thanks for specifying the area
Resolved

**Comment by btkcodedev** [comment]
Resolved

File: airbyte-integrations/connectors/source-twilio/acceptance-test-config.yml [hunk 9]
```diff
@@ -31,6 +31,8 @@ acceptance_tests:
             bypass_reason: "stream not filled yet"
           - name: "conference_participants"
             bypass_reason: "stream not filled yet"
+          - name: "conversation_participants"
+            bypass_reason: "stream not filled yet"
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-twilio/acceptance-test-config.yml:35]: please remove this. you're adding a new stream and we must validate it is working properly.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-twilio/source_twilio/streams.py [hunk 10]
```diff
@@ -363,6 +364,21 @@ class Keys(TwilioNestedStream):
 
     parent_stream = Accounts
 
+class ConversationParticipants(IncrementalTwilioStream, TwilioNestedStream):
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-twilio/source_twilio/streams.py:367]: This endpoint doesn't allow you to filter data based on dates so it isn't incremental.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

**Comment by btkcodedev** [comment]
Thanks for the comment :) Fixed

**Comment by btkcodedev** [comment]
But why this `Add labels to github PRs based on context / Add Labels to PRs. Safe to Merge on fail` always fails? :(

File: airbyte-integrations/connectors/source-twilio/source_twilio/schemas/conversation_participants.json [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+        },
+        "proxy_address": {
+          "type": ["null", "string"]
+        }
+      }
+    },
+    "role_sid": {
+      "type": ["null", "string"]
+    },
+    "date_created": {
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-twilio/source_twilio/schemas/conversation_participants.json:42]: This is not date-time? Please confirm the correct data format. As for date_updated

File: airbyte-integrations/connectors/source-twilio/source_twilio/schemas/conversation_participants.json [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    "attributes": {
+      "type": "object",
+      "properties": {
+        "role": {
+          "type": ["null", "string"]
+        }
+      }
+    },
+    "messaging_binding": {
+      "type": "object",
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-twilio/source_twilio/schemas/conversation_participants.json:26]: Are you going to always expect a message binding? Maybe is better ["null", "object"]

File: airbyte-integrations/connectors/source-twilio/source_twilio/schemas/conversation_participants.json [hunk 13]
```diff
@@ -0,0 +1,58 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "type": "object",
+  "properties": {
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-twilio/source_twilio/schemas/conversation_participants.json:4]: Missing additionalProperties.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
Left some comments.

**Comment by btkcodedev** [comment]
Update: Parent stream merged, Resumed work

**Comment by btkcodedev** [comment]
Ready for review :)

**Review by marcosmarxm — APPROVED** [review]
Thanks @btkcodedev I'm having problems with CI right now. I'll publish the connector on Monday.

File: airbyte-integrations/connectors/source-twilio/source_twilio/schemas/conversation_participants.json [hunk 14]
```diff
@@ -0,0 +1,61 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
```
Comment (artem1205) [inline airbyte-integrations/connectors/source-twilio/source_twilio/schemas/conversation_participants.json:2]: change $schema to https://json-schema.org/draft-07/schema#

**Review by artem1205 — CHANGES_REQUESTED** [review]

**Review by artem1205 — APPROVED** [review]

**Pull request merged**